lsm: disambiguate between empty database and applied seqno of 0

rockwotj opened the pull request:
Currently 0 was conflated with empty db and that can make replay
hard/annoying.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes


* none

rockwotj committed:
lsm: disambiguate between empty database and applied seqno of 0

Currently 0 was conflated with empty db and that can make replay
hard/annoying.

rockwotj commented on src/v/lsm/db/impl.cc:
```diff
@@ -528,14 +539,20 @@ ss::future<> impl::remove_obsolete_files() {
-      .or_else([this] { return _imm ? (*_imm)->last_seqno() : std::nullopt; })
-      .value_or(max_persisted_seqno());
+std::optional<internal::sequence_number> impl::max_applied_seqno() const {
+    if (auto seqno = _mem->last_seqno()) {
+        return seqno;
+    }
+    if (_imm) {
+        if (auto seqno = (*_imm)->last_seqno()) {
+            return seqno;
+        }
+    }
+    return max_persisted_seqno();
```
Disagree, you have poor taste robot!

rockwotj committed:
lsm: add epoch to the manifest

Just to make things explicit and allows you to know when reading the
manifest what version you read at.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--99906ba3-272e-4367-bc52-b00867a96def-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78083">build#78083</a></summary>


| test_class                 | test_method                 | test_arguments                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                         |
|:---------------------------|:----------------------------|:---------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeClusterRestoreTest | test_restore_partition_spec | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/78083#019b2f96-593d-4f47-acbe-e51ead4c53f7 | FLAKY         | 49/51    | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.0032, p0=0.1481, reject_threshold=0.0100. adj_baseline=0.0096, p1=0.9170, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_restore_partition_spec |
| MountUnmountIcebergTest    | test_simple_remount         | {"cloud_storage_type": 1}                                | integration | https://buildkite.com/redpanda/redpanda/builds/78083#019b2f96-5933-4d52-b5ed-064cfd6d8ba3 | FLAKY         | 7/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1707, p0=0.2362, reject_threshold=0.0100. adj_baseline=0.4297, p1=0.3108, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount            |
</details>
<!--/body-->

dotnwat reviewed (commented):
does the ambiguity only occur on brand new databases, or also on non-empty databases that become empty? if the former, is the solution to effectively create a formatting step which lays some initial state down on disk when a new db is created?

rockwotj commented:
> does the ambiguity only occur on brand new databases, or also on non-empty databases that become empty? if the former, is the solution to effectively create a formatting step which lays some initial state down on disk when a new db is created?

Only brand new databases. Becoming empty does not change the sequence number, which is effectively the offset applied to the database from the raft log

rockwotj committed:
lsm/db: add a snapshot list class

This class keeps trace of which explicit snapshots are held by users.

LevelDB link: https://github.com/google/leveldb/blob/main/db/snapshot.h

andrwng commented on src/v/lsm/lsm.cc:
```diff
@@ -163,12 +163,14 @@ ss::future<database> database::open(options opts, io::persistence p) {
-    return seqno_cast(_impl->max_persisted_seqno());
+std::optional<model::offset> database::max_persisted_offset() const {
+    return _impl->max_persisted_seqno().transform(
+      [](auto seqno) { return seqno_cast(seqno); });
 }
 
-model::offset database::max_applied_offset() const {
-    return seqno_cast(_impl->max_applied_seqno());
+std::optional<model::offset> database::max_applied_offset() const {
+    return _impl->max_applied_seqno().transform(
+      [](auto seqno) { return seqno_cast(seqno); });
```
BTW how do you feel about exposing seqno as the public interface? model::offset is nice in that it emphasizes being monotonically increasing, but it does feel kind of awkward/limiting if the Raft log offset isn't the exact source of these values.

rockwotj replied:
yeah I remember you mentioned this, I'm happy to push the sequence number concept to the public API. Let me do in a followup

dotnwat commented:
> > does the ambiguity only occur on brand new databases, or also on non-empty databases that become empty? if the former, is the solution to effectively create a formatting step which lays some initial state down on disk when a new db is created?
> 
> Only brand new databases. Becoming empty does not change the sequence number, which is effectively the offset applied to the database from the raft log

Got it. So is the solution in this PR something akin to formatting the db directory with initial state? Because that is the moment where there is no such ambiguity.

rockwotj commented:
> Got it. So is the solution in this PR something akin to formatting the db directory with initial state? Because that is the moment where there is no such ambiguity.

I'm not really sure what you mean by "formatting the db directory with initial state"?

Here's the concrete example where I hit this:

1. Open a database
2. We need to replay the log from what was just in memtables
3. ask the database what the last peristed offset is, and it returns 0

Does that mean offset 0 was persisted? Or does that mean we need to start from the beginning of the log? We have different mechanisms to represent empty logs in kafka land (hwm which is the next offset in the log not the last inclusive record's offset), but we don't have that here. Using nullopt removes the ambiguity, we could also change the API to be "next_seqno" but that opens up to all the inclusive/exclusive/off by one errors we have today with offsets in the kafka world.

dotnwat commented:
> > Got it. So is the solution in this PR something akin to formatting the db directory with initial state? Because that is the moment where there is no such ambiguity.
> 
> I'm not really sure what you mean by "formatting the db directory with initial state"?
> 
> Here's the concrete example where I hit this:
> 
> 1. Open a database
> 2. We need to replay the log from what was just in memtables
> 3. ask the database what the last peristed offset is, and it returns 0
> 
> Does that mean offset 0 was persisted? Or does that mean we need to start from the beginning of the log? We have different mechanisms to represent empty logs in kafka land (hwm which is the next offset in the log not the last inclusive record's offset), but we don't have that here. Using nullopt removes the ambiguity, we could also change the API to be "next_seqno" but that opens up to all the inclusive/exclusive/off by one errors we have today with offsets in the kafka world.

Since this is only a situation that arises with a brand new database, it suggests that brand-new is a state which can be used for disambiguation. Further, if db creation is a distinct operation (distinct from create-on-open semantics) then it follows that db creation can mark the precise situation in its on disk state. No need to derive the answer.
